rtrlib/aspa/aspa_array/aspa_array.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * This file is part of RTRlib.
+ *
+ * This file is subject to the terms and conditions of the MIT license.
+ * See the file LICENSE in the top level directory for more details.
+ *
+ * Website: http://rtrlib.realmv6.org/
+ */
+
+#include "aspa_array.h"
+
+#include "rtrlib/aspa/aspa_private.h"
+#include "rtrlib/lib/alloc_utils_private.h"
+#include "rtrlib/rtr/rtr.h"
+
+// MARK: - Initialization & Deinitialization
+
+enum aspa_status aspa_array_create(struct aspa_array **array_ptr)
+{
+	const size_t default_initial_size = 128;
+
+	// allocation the chunk of memory of the provider as numbers
+	struct aspa_record *data_field = lrtr_malloc(sizeof(struct aspa_record) * default_initial_size);
+
+	// malloc failed so returning an error
+	if (!data_field)
+		return ASPA_ERROR;
+
+	// allocating the aspa_record itself
+	struct aspa_array *array = lrtr_malloc(sizeof(struct aspa_array));
+
+	// malloc for aspa_record failed hence we return an error
+	if (!array) {
+		lrtr_free(data_field);
+		return ASPA_ERROR;
+	}
+
+	// initializing member variables of the aspa record
+	array->capacity = default_initial_size;
+	array->size = 0;
+	array->data = data_field;
+
+	// returning the array
+	*array_ptr = array;
+
+	return ASPA_SUCCESS;
+}
+
+void aspa_array_free(struct aspa_array *array, bool free_provider_arrays)
+{
+	// if the array is null just return
+	if (!array)
+		return;
+
+	if (array->data) {
+		if (free_provider_arrays) {
+			for (size_t i = 0; i < array->size; i++) {
+				if (array->data[i].provider_asns) {
+					lrtr_free(array->data[i].provider_asns);
+					array->data[i].provider_asns = NULL;
+				}
+			}
+		}
+
+		// freeing the data
+		lrtr_free(array->data);
+	}
+
+	// freeing the array itself
+	lrtr_free(array);
+}
+
+// MARK: - Manipulation
+
+static enum aspa_status aspa_array_reallocate(struct aspa_array *array)
```

Comment:
We can totally give the user a create function which takes the initial size of the array. But later one we want to use this array as a highy level data structure (comparable to `std::vector<>`) where we don't need to worry about resizing, that's why we built this abstraction.

Reply:
My assumption is that whenever more entries to the array are appended, the number of elements that are added is known beforehand. 
Therefore the array could be allocated with the exact size that is required.

This guarantees that:
1. Only 1 malloc operation is needed, with a dynamically resized array sometimes multiple ones will be done.
2. Only the exact amount of memory that is needed has been allocated

This is why a fixed size array seems superior in this usecase.
Am I missing something here?

Reply:
I added the `aspa_array_reserve` function which can pre-allocate memory and started using it at all places, where we know the size already before.  

(see last commit)

Reply:
great :tada:

Could you elaborate on in which situations the size of the array is not known before and why? 
It's not clear to me and it would take me much more type to figure it out on my own. :-)
Thanks!